kevm-pyk/src/kevm_pyk/__main__.py

```diff
@@ -325,38 +327,137 @@ def exec_foundry_prove(
-        _cfg_id, _cfg = _id_and_cfg
-        _claim = _kcfg_unproven_to_claim(_cfg)
+    def _write_cfg(_cfg: KCFG, _cfgpath: Path) -> None:
+        with open(_cfgpath, 'w') as cfgfile:
+            cfgfile.write(json.dumps(_cfg.to_dict()))
+            _LOGGER.info(f'Updated CFG file: {_cfgpath}')
+
+    def prove_it(_id_and_cfg: Tuple[str, Tuple[KCFG, Path]]) -> bool:
+        _cfg_id, (_cfg, _cfg_path) = _id_and_cfg
+        if len(_cfg.frontier) == 0:
+            return True
+        _init_node = _cfg.frontier[0]
+        _target_node = _cfg.get_unique_target()
+        _claim = KCFG.Edge(_init_node, _target_node, mlTop(), -1).to_claim()
         _claim_id = _cfg_id.replace('.', '-').replace('_', '-')
         ret, result = KProve_prove_claim(foundry, _claim, _claim_id, _LOGGER, depth=depth, lemmas=lemma_rules)
-        if minimize:
-            result = minimize_term(result)
-        print(f'Result for {_cfg_id}:\n{foundry.pretty_print(result)}\n')
-        return ret
+        _cfg.add_expanded(_init_node.id)
+        if is_top(result):
+            _cfg.create_edge(_cfg.get_unique_init().id, _cfg.get_unique_target().id, mlTop(), -1)
+            _LOGGER.info(f'Proof passed: {_cfg_id}')
+        else:
+            for result_state in flatten_label('#Or', result):
+                _cfg.add_expanded(_init_node.id)
+                new_node = _cfg.get_or_create_node(CTerm(result_state))
+                _cfg.create_edge(_cfg.get_unique_init().id, new_node.id, mlTop(), -1)
+                if minimize:
+                    result_state = minimize_term(result_state)
+                _LOGGER.error(f'Proof failed: {_cfg_id}\n{foundry.pretty_print(result_state)}')
+        _write_cfg(_cfg, _cfg_path)
+        failure_nodes = _cfg.frontier + _cfg.stuck
+        return len(failure_nodes) == 0
 
     with ProcessPool(ncpus=workers) as process_pool:
         results = process_pool.map(prove_it, kcfgs.items())
         process_pool.close()
 
-    failed_cfgs = [cid for ((cid, _), failed) in zip(kcfgs.items(), results) if failed]
-    if failed_cfgs:
-        print(f'Failed to prove KCFGs: {failed_cfgs}\n')
-    sys.exit(len(failed_cfgs))
+    failed = 0
+    for (cid, _), succeeded in zip(kcfgs.items(), results):
+        if succeeded:
+            print(f'PASSED: {cid}')
+        else:
+            print(f'FAILED: {cid}')
+            failed += 1
+    sys.exit(failed)
+
+
+def exec_foundry_show(
+    profile: bool,
+    foundry_out: Path,
+    test: str,
+    nodes: Iterable[str] = (),
+    node_deltas: Iterable[Tuple[str, str]] = (),
+    minimize: bool = True,
+    **kwargs: Any,
+) -> None:
+    definition_dir = foundry_out / 'kompiled'
+    use_directory = foundry_out / 'specs'
+    use_directory.mkdir(parents=True, exist_ok=True)
+    kcfgs_dir = foundry_out / 'kcfgs'
+    foundry = Foundry(definition_dir, profile=profile, use_directory=use_directory)
+    kcfg_file = kcfgs_dir / f'{test}.json'
```

Comment:
Consider extracting `definition_dir`, `use_dir`, `kcfg_dir`, `kcfg_file` as functions of a new module `config.py`, e.g.

```python3
def kcfg_file(out_dir: Path, test_name: str) -> Path:
    return kcfg_dir(out_dir) / f'{test_name}.json'
```

Reply:
You mean as helpers so we're not re-computing them over and again? Maybe they could be fields in the `KEVM` class so they're jsut set at function init?

Reply:
I took a quick look at this, and seems like more trouble than it's worth at the moment. Will consider it in my local tree, see if I can find a solution.

Reply:
> You mean as helpers so we're not re-computing them over and again

Re-computing the values is fine. The point is to have a single definition of these expressions so that e.g. the folder layout can be easily reconfigured.